### PR TITLE
BSIP35 mitigate rounding issue when matching orders

### PR DESCRIPTION
**Pull request by abitmore**
This is a subset of changes on the market engine, cherry-picked from #641, including changes related to #132, #184, #342 and [BSIP35: Mitigate Rounding Issue On Order Matching](https://github.com/bitshares/bsips/blob/master/bsip-0035.md).

The code is ready for review. In the meanwhile I'll finish the missing test cases, including:
- something-for-nothing issue
  - [x] matching two limit orders
  - [x] matching a limit order with a call order
  - [x] matching a settlement order with a call order
  - [x] globally settling
  - [x] settling after globally settled
- mitigate rounding issue
  - [x] matching two limit orders
  - [x] matching a limit order with a call order
  - [x] matching a settlement order with a call order
  - [x] globally settling
  - [x] settling after globally settled

### CODE REVIEW

File: libraries/chain/db_market.cpp [hunk 1]
```diff
@@ -585,11 +646,70 @@ asset database::match( const call_order_object& call,
...
+         {
+            // be here, call_pays has been rounded down
+
+            // be here, we should have: call_pays <= call_collateral
+
+            if( call_receives == settle.balance ) // the settle order will be completely filled, assuming we need to cull it
+               cull_settle_order = true;
+            // else do nothing, since we can't cull the settle order
+
+            call_receives = call_pays ^ match_price; // round up here to mitigate rounding issue (core-342)
```
Comment (pmconrad) [inline libraries/chain/db_market.cpp]: It is important to understand here that the newly rounded up call_receives cannot be greater than the old call_receives.

File: libraries/chain/asset_evaluator.cpp [hunk 2]
```diff
@@ -508,33 +508,54 @@ void_result asset_settle_evaluator::do_evaluate(const asset_settle_evaluator::op
...
+         else // TODO remove this warning after hard fork core-184
+            wlog( "Something for nothing issue (#184, variant F) occurred at block #${block}", ("block",d.head_block_num()) );
+      }
+
+      asset pays = op.amount;
+      if( op.amount.amount != mia_dyn.current_supply
+            && settled_amount.amount != 0
+            && d.get_dynamic_global_properties().next_maintenance_time > HARDFORK_CORE_342_TIME )
+      {
+         pays = settled_amount ^ bitasset.settlement_price;
```
[the remark this answers is not in the document]
Comment (abitmore) [reply libraries/chain/asset_evaluator.cpp]: Changed to a method: https://github.com/bitshares/bitshares-core/pull/830/commits/c6b28b2f0534f3a5710543a9d6f029a614432382.

File: libraries/chain/db_market.cpp (hunk 1 above)
Comment (abitmore) [reply libraries/chain/db_market.cpp]: Added your comment.

BTW I have a test case for this in the bsip38 PR, need to update atm due to the operator change though.

File: libraries/chain/db_update.cpp [hunk 3]
```diff
@@ -423,10 +429,28 @@ void database::clear_expired_orders()
...
-               settled += match(*itr, order, settlement_price, max_settlement, settlement_fill_price);
+               asset new_settled = match(*itr, order, settlement_price, max_settlement, settlement_fill_price);
+               if( maint_time > HARDFORK_CORE_184_TIME && new_settled.amount == 0 ) // unable to fill this settle order
+               {
+                  if( find_object( order_id ) ) // the settle order hasn't been cancelled
+                     current_asset_finished = true;
+                  break;
+               }
+               settled += new_settled;
+               // before hard fork core-342, if new_settled > 0, we'll have:
```
[the remark this answers is not in the document]
Comment (abitmore) [reply libraries/chain/db_update.cpp]: Done.

**Comment by abitmore** [comment]
@pmconrad all done. Please review again.

**Comment by pmconrad** [comment]
I'm missing something-for-nothing variant C-1 in the test output. Is that also just a theoretical possibility, like D-1/D-2, or was it overlooked?

**Comment by abitmore** [comment]
@pmconrad added more tests to cover C-1 and cull-settle logic.

File: libraries/chain/db_market.cpp [hunk 4]
```diff
@@ -539,37 +580,57 @@ int database::match( const limit_order_object& bid, const call_order_object& ask
...
       order_receives  = usd_for_sale * match_price; // round down here, in favor of call order
-      call_pays       = order_receives;
-      order_pays      = usd_for_sale;
 
-      filled_limit    = true;
-      filled_call     = ( usd_to_buy == usd_for_sale );
+      // Be here, it's possible that taker is paying something for nothing due to partially filled in last loop.
+      // In this case, we see it as filled and cancel it later
+      // TODO remove hardfork check when we're sure it's always true (but keep the zero amount check)
+      if( order_receives.amount == 0 && maint_time > HARDFORK_CORE_184_TIME )
```
Comment (pmconrad) [inline libraries/chain/db_market.cpp]: Perhaps add an ilog() here so we can later check if the case is relevant.

File: libraries/chain/db_market.cpp [hunk 5]
```diff
@@ -539,37 +580,57 @@ int database::match( const limit_order_object& bid, const call_order_object& ask
...
+         return 1;
+
+      if( before_core_hardfork_342 ) // TODO remove this "if" when we're sure it's always false (keep the code in else)
```
Comment (pmconrad) [inline libraries/chain/db_market.cpp]: Perhaps add an ilog() here so we can later check if the case is relevant.

File: libraries/chain/db_market.cpp [hunk 6]
```diff
@@ -539,37 +580,57 @@ int database::match( const limit_order_object& bid, const call_order_object& ask
...
-      order_receives = usd_to_buy * match_price; // round down here, in favor of call order
-      call_pays      = order_receives;
-      order_pays     = usd_to_buy;
-
-      filled_call    = true;
+      if( before_core_hardfork_342 ) // TODO remove this "if" when we're sure it's always false (keep the code in else)
+      {
+         order_receives = usd_to_buy * match_price; // round down here, in favor of call order
+         // TODO remove hardfork check when we're sure it's always true (but keep the zero amount check)
+         if( order_receives.amount == 0 && maint_time > HARDFORK_CORE_184_TIME )
```
Comment (pmconrad) [inline libraries/chain/db_market.cpp]: Perhaps add an ilog() here so we can later check if the case is relevant.

**Pull request merged**